### PR TITLE
Extract pov rejects in Wikipedia

### PR DESCRIPTION
**Pull request by vegetable68**
This PR:

- Separates the functionality of sentence level diff with the main dataflow pipeline.
- Adds unit test to this functionality.

### CODE REVIEW

File: extract_pov_edits/dataflow_main.py [hunk 1]
```diff
@@ -236,7 +177,9 @@ def process(self, element, bucket, ingestFrom):
     if ingestFrom == 'local':
        input_stream = chunk_name
     else:
-       os.system("gsutil -m cp %s %s" % (path.join('gs://', bucket, chunk_name), chunk_name))
+       cmd = "gsutil -m cp %s %s" % (path.join('gs://', bucket, chunk_name), chunk_name)
+       if WEXITSTATUS(os.system()) != 0:
+         raise Excetion("GSUTIL COPY Error")
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: Excetion ? - also capture the os.system() exit status so you can log it.

File: extract_pov_edits/dataflow_main.py [hunk 2]
```diff
@@ -367,4 +316,14 @@ def directory(mirror):
            sections.append(obj.name[obj.name.rfind('/') + 1:])
   if known_args.ingestFrom == 'local':
      sections = [known_args.localStorage]
-  run(known_args, pipeline_args, sections)
+  if known_args.start is not None:
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: just:
if know_args.start:

but prob
start = known_args.start or 0
end = known_args.end or 0

File: extract_pov_edits/ingest_utils/process.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    if inc + ind < length and changes[inc + ind]:
+      return True
+    if ind - inc >= 0 and changes[ind - inc]:
+      return True
+  return False
+
+
+def diff(sents, cur_sents, rev_id):
+  """Computes diff at sentence level."""
+  sents_old = [s for s in cur_sents.keys()]
```
Comment (sorensenjs) [inline extract_pov_edits/ingest_utils/process.py]: this is just sents_old = cur_sents.keys()
I told you this the previous time, no? Seems familiar.

File: extract_pov_edits/ingest_utils/process.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+def split(text):
+  """Split the paragraph using !?. and line break."""
+  sentence_break = '!?.\n'
+  length = len(text)
+  start = 0
+  for ind, ch in enumerate(text):
+    if (ind == length - 1) or (ch in sentence_break
+                               and (text[ind+1] not in sentence_break)):
+       if text[start:ind+1].strip() != '':
```
Comment (sorensenjs) [inline extract_pov_edits/ingest_utils/process.py]: just if text[start:ind + 1].strip():
but you're yielding it so extract to local variable.

File: extract_pov_edits/ingest_utils/process.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def split(text):
+  """Split the paragraph using !?. and line break."""
+  sentence_break = '!?.\n'
+  length = len(text)
+  start = 0
+  for ind, ch in enumerate(text):
+    if (ind == length - 1) or (ch in sentence_break
+                               and (text[ind+1] not in sentence_break)):
```
Comment (sorensenjs) [inline extract_pov_edits/ingest_utils/process.py]: spaces around operators (+)

File: extract_pov_edits/ingest_utils/process.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+          yield text[start:ind+1].strip()
+       start = ind+1
+
+
+def process(content, cur_sents):
+  """Main Processing Point."""
+  content['text'] = format_clean(content['text'])
+  sents = []
+  error = False
+  for component in split(content['text']):
```
Comment (sorensenjs) [inline extract_pov_edits/ingest_utils/process.py]: this is an unnecessary hash lookup, hold onto format_clean(content['text']))
maybe python JIT is smart, but I wouldn't count on it.

File: extract_pov_edits/process_test.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          content = json.loads(line)
+          ans, error = process(content, cur_sents)
+          anses.append(ans)
+    with open(ans_file, 'r') as f:
+      standard = f.read()
+    self.assertEqual(json.dumps(anses), standard)
+
+  def test_large_revs(self):
+    content = {'rev_id': 'placeholder'}
+    content['text'] = 'w' * 500000 + '. Haha there\'s a sentence.'
```
Comment (sorensenjs) [inline extract_pov_edits/process_test.py]: LOL, I was thinking of filing a bug against NLTK for this.

File: extract_pov_edits/ingest_utils/process.py (hunk 3 above)
Comment (vegetable68) [reply extract_pov_edits/ingest_utils/process.py]: Sorry! Fixed now.

File: extract_pov_edits/dataflow_main.py (hunk 2 above)
Comment (vegetable68) [reply extract_pov_edits/dataflow_main.py]: Yeah that's the kind of confusion I want to avoid here, they can be 0.

Comment (sorensenjs) [reply extract_pov_edits/dataflow_main.py]: OK

**Pull request merged**